signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+
+    def reset(self, data=None):
+        if data is None:
+            data = []
+        if isinstance(data, Sequence) and not isinstance(data, str):
+            with self._suspend_sync():
+                backup = copy(self._data)
+                try:
+                    for i in range(min(len(self), len(data))):
+                        if data[i] == self._data[i]:
+                            continue
+                        try:
+                            self._data[i].reset(data[i])
+                            continue
+                        except (ValueError, AttributeError):
+                            pass
+                        self._data[i] = self.from_base(data[i])
+                    if len(self._data) > len(data):
+                        self._data[:len(data)]
+                    else:
+                        self.extend(data[len(self)])
+                    self.sync()
+                except BaseException:  # rollback
+                    self._data = backup
+                    raise
+        else:
+            raise ValueError("The data must be a sequence or None.")
+
+    def __delitem__(self, item):
+        self.load()
+        del self._data[item]
+        self.sync()
+
+    def __setitem__(self, key, value):
+        self.load()
+        with self._suspend_sync():
+            self._data[key] = self.from_base(value, self)
+        self.sync()
+
+    def __getitem__(self, key):
+        self.load()
+        return self._data[key]
+
+    def __iter__(self):
+        self.load()
+        return iter(self._data)
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def __reversed__(self):
+        self.load()
+        return reversed(self._data)
+
+    def __iadd__(self, iterable):
+        self.load()
+        self._data += [self.from_base(value, self) for value in iterable]
+        self.sync()
+
+    def insert(self, index, item):
+        self.load()
+        with self._suspend_sync():
+            self._data.insert(index, self.from_base(item, self))
+        self.sync()
+
+    def append(self, item):
+        self.load()
+        with self._suspend_sync():
+            self._data.append(self.from_base(item, self))
+        self.sync()
+
+    def extend(self, iterable):
+        self.load()
+        with self._suspend_sync():
+            self._data.extend([self.from_base(value, self) for value in iterable])
+        self.sync()
+
+    def remove(self, item):
+        self.load()
+        with self._suspend_sync():
+            self._data.remove(self.from_base(item, self))
+        self.sync()
+
+    def clear(self):
+        self._data = []
+        self.sync()
+
+
+class JSONCollection(SyncedCollection):
+
+    def __init__(self, filename, write_concern=False):
+        self._filename = os.path.realpath(filename)
+        self._write_concern = write_concern
+
+    def _load_from_disk(self):
+        try:
+            with open(self._filename, 'rb') as file:
```

Comment:
Is there a reason to prefer `'rb'` and `decode()` versus just using `'r'`? (I have not tested this.)

Reply:
I think it's slightly more performant and explicit.

Reply:
@csadorf is probably right about performance. I will resolve this comment.